### PR TITLE
Bug 1966586: 200 OK returned when setting invalid BaseDNSdomain via API

### PR DESCRIPTION
**Pull request by sagidayan**
# Description

Updated `ValidateDNSName`.                                                                 
The function was validating against `apps.<clusterName>.<baseDNSDomain>` .
This is why via the API `baseDNSDomain=example` is responding with 200. because it actually validates `apps.<cluster_name>.example`.

The update adds a validation on the baseDNS itself in addition to the validations we already have.

Updated subsytem tests to ensure validation

# What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

# How was this code tested?

Please, select one or more if needed:

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

It's recommended to take a few extra minutes to provide more information about
how this code was tested. Here are some questions that may be worth answering:

- Should this PR be tested by the reviewer?
- Is this PR relying on CI for an e2e test run?
- Should this PR be tested in a specific environment?
- Any logs, screenshots, etc that can help with the review process?

# Assignees

Please, add one or two reviewers that could help review this PR. Use `/assign` if you want to assign
this PR directly to someone.

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -3811,6 +3811,7 @@ definitions:
       base_dns_domain:
         type: string
         description: Base domain of the cluster. All DNS records must be sub-domains of this base and include the cluster name.
+        pattern: '^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}$'
```
Comment (michaellevy101) [inline swagger.yaml:3814]: Where did you find this pattern?
is the error message indicative to the user?

**Review by michaellevy101 — COMMENTED** [review]
What about kubeapi? (You can create a cluster from the kubeapi as well)
What about testing?

File: swagger.yaml (hunk 1 above)
Comment (sagidayan) [reply swagger.yaml:3814]: Pattern was found In the UI validation (https://github.com/openshift-assisted/assisted-ui-lib/blob/76d85794aa565e98fb3820df879baf17525fc9a2/src/components/ui/formik/validationSchemas.ts#L15). So it uses same validation.

Regarding the error message, It uses the swagger default message, same as any other patterned parameter

**Comment by sagidayan** [comment]
@michaellevy101 
> What about kubeapi? (You can create a cluster from the kubeapi as well)

We have two epics that will tackle kubeapi validations.
 - MGMT-5247
 - MGMT-6677


> What about testing?
Added

File: subsystem/cluster_test.go [hunk 2]
```diff
@@ -821,6 +821,55 @@ var _ = Describe("cluster install - DHCP", func() {
...
+				ClusterNetworkHostPrefix: 23,
+				Name:                     swag.String("test-cluster"),
+				OpenshiftVersion:         swag.String(openshiftVersion),
+				PullSecret:               swag.String(pullSecret),
+				ServiceNetworkCidr:       &serviceCIDR,
+				SSHPublicKey:             sshPublicKey,
+			},
+		})
+	}
+	It("Should thow en error with invalid BaseDNS 'example'", func() {
```
Comment (slaviered) [inline subsystem/cluster_test.go:847]: 1. Spelling: "throw an error"
2. Use loop of tests instead a copy paste of the same test logic
3. Please use more descriptive test names. for example: "Illegal character in domain name", "too short domain name" etc.

File: subsystem/cluster_test.go [hunk 3]
```diff
@@ -821,6 +821,55 @@ var _ = Describe("cluster install - DHCP", func() {
 	})
 })
 
+var _ = Describe("Create cluster with BaseDNS", func() {
```
Comment (slaviered) [inline subsystem/cluster_test.go:824]: Use more descriptive name: such validation of baseDNS during create cluster

File: swagger.yaml [hunk 4]
```diff
@@ -4101,6 +4103,7 @@ definitions:
...
```
Comment (slaviered) [inline swagger.yaml:4106]: Please consult Avishai if we do not need to extract base_dns_domain to the description part and use ref here

File: subsystem/cluster_test.go (hunk 2 above)
Comment (sagidayan) [reply subsystem/cluster_test.go]: 👍🏽 Thanks for the input!

File: subsystem/cluster_test.go (hunk 3 above)
Comment (sagidayan) [reply subsystem/cluster_test.go]: 👍🏽

File: swagger.yaml (hunk 4 above)
Comment (sagidayan) [reply swagger.yaml]: We decided to remove the swagger changes altogether.
Please see description.

File: internal/dns/dns.go [hunk 5]
```diff
@@ -192,13 +192,16 @@ func (h *handler) GetDNSDomain(clusterName, baseDNSDomainName string) (*DNSDomai
...
+// ValidateDNSName checks baseDNSDomainName and the combination of cluster name and base DNS domain
 // leaves enough room for automatically added domain names,
 // e.g. "alertmanager-main-openshift-monitoring.apps.test-infra-cluster-assisted-installer.example.com").
 // The max total length of a domain name is 255 bytes, including the dots. An individual label can be
 // up to 63 bytes. A single char may occupy more than one byte in Internationalized Domain Names (IDNs).
 func (h *handler) ValidateDNSName(clusterName, baseDNSDomainName string) error {
 	appsDomainNameSuffix := fmt.Sprintf(appsDomainNameFormat, clusterName, baseDNSDomainName)
+	if err := validations.ValidateDomainNameFormat(baseDNSDomainName); err != nil {
+		return err
+	}
```
Comment (ori-amizur) [inline internal/dns/dns.go:204]: Is there a reason to validate **appsDomainNameFormat** now ?

Comment (sagidayan) [reply internal/dns/dns.go:204]: You are right. The format of `appsDomainNameSuffix` should be valid if `clusterName` is valid and the `baseDNS` is valid.
The only thing to validate is `appsDomainNameSuffix`'s is length. 
Ill update.

**Pull request merged**